### PR TITLE
Authority record digital object linking, refs #12650

### PR DESCRIPTION
**Pull request by sbreker**
Allow authority records to link a single digital object. This functionality matches the current "Link Digital Object" functionality currently available for descriptions.  This feature includes the ability to set the alt text for the digital object - the alt text field can be set on the digital object edit page.

Access control behaves differently from information objects. There are no PREMIS rules for authorities.  Access to these digital objects is controlled via the existing authority record ACL controls. Editing the digital object requires authority update permissions etc.

- Must have read privileges to view
- PDF master is always accessible to view as with Descriptions
- Create permission is needed to create an actor
- Update privileges are needed to add or update an actor linked digital object as this is technically updating the actor
- Delete privileges are needed to delete

### CODE REVIEW

File: plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php [hunk 1]
```diff
@@ -259,11 +308,28 @@ public function serialize()
       $serialized['directSubjects'][] = $item->id;
     }
 
+<<<<<<< HEAD
```
Comment (mcantelon) [inline plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php]: Merge artefact.

File: plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php [hunk 2]
```diff
@@ -259,11 +308,28 @@ public function serialize()
...
+      $serialized['digitalObject']['filename'] = $this->filename;
+      $serialized['digitalObject']['thumbnailPath'] = $this->getThumbnailPath();
+      $serialized['digitalObject']['digitalObjectAltText'] = $this->getDigitalObjectAltText();
+
+      $serialized['hasDigitalObject'] = true;
+    }
+    else
+    {
+      $serialized['hasDigitalObject'] = false;
+>>>>>>> Auth DO update ES Actor indexing, refs #12650
```
Comment (mcantelon) [inline plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php]: Merge artefact.

File: plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php [hunk 3]
```diff
@@ -259,11 +308,28 @@ public function serialize()
       $serialized['directSubjects'][] = $item->id;
     }
 
+<<<<<<< HEAD
     // Maintenance notes
     $sql = 'SELECT id, source_culture FROM '.QubitNote::TABLE_NAME.' WHERE object_id = ? AND type_id = ?';
     foreach (QubitPdo::fetchAll($sql, array($this->id, QubitTerm::MAINTENANCE_NOTE_ID)) as $item)
     {
       $serialized['maintenanceNotes'][] = arElasticSearchNote::serialize($item);
+=======
```
Comment (mcantelon) [inline plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php]: Merge artefact.

File: apps/qubit/modules/actor/templates/_searchResult.php [hunk 4]
```diff
@@ -1,6 +1,27 @@
 <?php use_helper('Text') ?>
 
-<article class="search-result">
+<?php if (isset($doc['hasDigitalObject']) && true === $doc['hasDigitalObject']): ?>
```
Comment (mcantelon) [inline apps/qubit/modules/actor/templates/_searchResult.php]: Could use !empty().

File: apps/qubit/modules/actor/templates/_searchResult.php [hunk 5]
```diff
@@ -1,6 +1,27 @@
...
+  <article class="search-result has-preview">
+<?php else: ?>
+  <article class="search-result">
+<?php endif; ?>
+
+  <?php if (isset($doc['hasDigitalObject']) && true === $doc['hasDigitalObject']): ?>
```
Comment (mcantelon) [inline apps/qubit/modules/actor/templates/_searchResult.php]: Could use !empty().

File: apps/qubit/modules/digitalobject/actions/deleteAction.class.php [hunk 6]
```diff
@@ -59,16 +58,27 @@ public function execute($request)
...
+
+      // Redirect to edit page for parent Object
       if (isset($parent))
       {
         $this->redirect(array($parent, 'module' => 'digitalobject', 'action' => 'edit'));
       }
       else
       {
-        $this->redirect(array($this->object, 'module' => 'informationobject'));
+        if ($this->object instanceOf QubitInformationObject)
```
Comment (mcantelon) [inline apps/qubit/modules/digitalobject/actions/deleteAction.class.php:74]: Could make this briefer via something like $redirectModule = ($this->object instanceOf QubitInformationObject) ? 'informationobject' : 'actor'

File: lib/model/QubitActor.php [hunk 7]
```diff
@@ -648,4 +669,135 @@ public static function setTermRelationByName($name, $options = array())
...
+  *****************/
+  /**
+   * Wrapper for QubitDigitalObject::importFromUri() method
+   *
+   * @param array $uris URIs of remote files
+   * @return QubitActor $this
+   *
+   * @TODO allow for different usage types
+   */
+  public function importDigitalObjectFromUri($uris, &$errors)
```
Comment (mcantelon) [inline lib/model/QubitActor.php:684]: Given this code (and importDigitalObjectFromBase64, etc.) is very similar to QubitInformationObject::importDigitalObjectFromUri maybe we can make a Redmine issue, for the future, to make this shared (in a child class of QubitObject or somewhere).

File: plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php [hunk 8]
```diff
@@ -322,4 +374,18 @@ protected function getDirectlyRelatedTerms($typeId)
 
     return self::$statements['relatedTerms']->fetchAll(PDO::FETCH_OBJ);
   }
+
+  protected function getProperty($name)
```
Comment (mcantelon) [inline plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php:378]: We should probably make an issue, as this is shared with arElasticSearchInformationObjectPdo, to refactor this so the code's shared.

**Review by mcantelon — CHANGES_REQUESTED** [review]
I've notes some small changes that could be made, and some refactoring we might want to do in the future, but all in all darn impressive!

File: plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php (hunk 1 above)
Comment (sbreker) [reply plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php]: Fixed in commit: Auth DO fix es actor pdo class, refs #12650

File: plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php (hunk 2 above)
Comment (sbreker) [reply plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php]: Fixed in commit: Auth DO fix es actor pdo class, refs #12650

File: plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php (hunk 3 above)
Comment (sbreker) [reply plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php]: Fixed in commit: Auth DO fix es actor pdo class, refs #12650

File: apps/qubit/modules/actor/templates/_searchResult.php (hunk 4 above)
Comment (sbreker) [reply apps/qubit/modules/actor/templates/_searchResult.php]: Fixed.  See commit: Auth DO search template changes, refs #12650

File: apps/qubit/modules/actor/templates/_searchResult.php (hunk 5 above)
Comment (sbreker) [reply apps/qubit/modules/actor/templates/_searchResult.php]: Fixed. See commit: Auth DO search template changes, refs #12650

File: apps/qubit/modules/digitalobject/actions/deleteAction.class.php (hunk 6 above)
Comment (sbreker) [reply apps/qubit/modules/digitalobject/actions/deleteAction.class.php:74]: This change to link to object leaves the door open for attaching digital objects to other object types. Changing the code as suggested leaves an assumption that if it's not infoObj, then it's an actor which _could_ sneak in as a future regression? I'd lean towards the long version as implemented I think.  Thoughts?

Comment (mcantelon) [reply apps/qubit/modules/digitalobject/actions/deleteAction.class.php:74]: True... good thinking.

**Review by mcantelon — APPROVED** [review]
Changes look good!

File: plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php (hunk 8 above)
Comment (sbreker) [reply plugins/arElasticSearchPlugin/lib/model/arElasticSearchActorPdo.class.php:378]: I agree.  Will see that tickets are created.

File: lib/model/QubitActor.php (hunk 7 above)
Comment (sbreker) [reply lib/model/QubitActor.php:684]: I agree.  Will see that tickets are created.

**Pull request closed without merging**